### PR TITLE
Add typescript support

### LINKED ISSUE #107: Typescript migration
## Problem that this feature solves

Higher code quality and minimal errors

## Feature description

Migrate all code from Javascript to Typescript

## Additional comments

Please check the following resources:
  * https://v3.vuejs.org/guide/typescript-support.html

### PR DESCRIPTION
**Pull request by aleesaan**
## What changes have been made? 

- Added support for typescript, disabling `noImplicitAny` for now to keep changes minimal
- Updated some dependencies
- Created separate config files to have a better separation of concerns
- Fixed linters as we were getting conflicts, now the code is shiny ✨ 
- Refactored store modules with static data into consts
- Moved tests files to match the folder structure suggested by the vue cli
- Left the components tests in `js` as they would require additional changes, and since we should refactor them anyway (to split them into invididual files), we can migrate them to `ts` then.

Note that most of the changes are automatic lint fixes done by the cli, especially in component files.

### Issues resolved by these changes
Resolves #107

## Additional comments
This is a follow up case to @RainnyX's [PR](https://github.com/xrnl/aginun/pull/115), which had some issues with styling, and I couldn't get it to work. I had tried improving some things in a separate [PR](https://github.com/RainnyX/aginun/pull/1), but I just couldn't solve the styling issues: I slowly downgraded the various packages, but when it still didn't work, I ended up starting from scratch and added our original changes one at the time, without upgrading vuetify-related things. We can look into upgrading vuetify (and figure out what breaks) in a later case.

### CODE REVIEW

File: src/store/modules/index.js [hunk 1]
```diff
@@ -1,88 +0,0 @@
-// Functionality obtained from:
-// https://github.com/chrisvfritz/vue-enterprise-boilerplate/blob/master/src/state/modules/index.js
```
Comment (SamLubbers) [inline src/store/modules/index.js:2]: I'm very happy you've deleted this horrible file that automatically loaded modules :grin: It was extremely long and hard to understand. The current manual loading of modules in `store/index.ts` is much much better :sparkles:

**Review by SamLubbers — CHANGES_REQUESTED** [review]
@aleesaan and @RainnyX this change is AMAZING :heart_eyes: :heart_eyes: Richard thank you for building the foundation and Ale thank you for fixing all the issues, painstakingly installing packages one by one, and making such a clear pull request :green_heart: 

I've just finished the first part of the review, which consisted of testing the application and checking all vue and configuration files. I'm already sharing my feedback so you can implement which I do the second part of the review: checking the new typescript files.

For the configuration files I've left all my comments in the code. Here is the rest of my feedback:

## Issues

I need to admit that I was surprised the application still worked so well after all these changes :laughing: I've only encountered one issue

## Unable to create / edit roles

When I create or edit a role, I get the following error and the role doesn't create/edit: `Uncaught (in promise) Error: GraphQL error: field "timeCommitments" not found in type: 'role_insert_input'`

## Remaining functionality

I think it is really good that you're taking  a gradual approach to migrating to typescript. Is it an idea to make issues for all the parts that we still need to migrate to typescript? That way we can easily divide this work between you, Richard, me or anyone else who'd like to help.

File: .prettierrc.json [hunk 2]
```diff
@@ -0,0 +1,4 @@
+{
+  "printWidth": 120,
+  "arrowParens": "always"
+}
```
[the remark this answers is not in the document]
Comment (aleesaan) [reply .prettierrc.json:4]: - In typescript we'll need to declare types for the arguments, so we'll always need brackets around function params.
- I only changed it cause I think 80 is a bit too restrictive, but I'll just drop `printWidth` altogether and just use the default (`80`).
- I only dropped them because they're already the default values.

**Comment by aleesaan** [comment]
Thanks for the thorough testing and comments @SamLubbers! I've already fixed the error you mentioned about creating/editing roles 💪   
Next I'll look into the specific comments about the configs, but just as a general answer: most of the changes I did were to match the configurations generated by the vue-cli, which I think is good to follow! Still, I'll look into your comments in detail tomorrow so I can see more specifically ✌️

File: .eslintrc.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+module.exports = {
+  root: true,
+  env: {
+    node: true
```
[the remark this answers is not in the document]
Comment (aleesaan) [reply .eslintrc.js:4]: I followed the configuration given by the vue cli, and they didn't have these. From my testing the `window` object is still working fine though, or did you notice any issue?

File: .eslintrc.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  },
+  extends: ["plugin:vue/essential", "@vue/airbnb", "@vue/typescript", "plugin:prettier/recommended"],
```
[the remark this answers is not in the document]
Comment (aleesaan) [reply .eslintrc.js:6]: I had added the `airbnb` plugin as airbnb is considered a [good standard](https://github.com/airbnb/javascript) for codestyle, and because it's the first lint configuration suggested by the cli. I now removed it and went with the 'prettier' lint configuration from the cli; I still have to push, but this is how it looks:
```
"plugin:vue/essential",
"eslint:recommended",
"@vue/typescript/recommended",
"@vue/prettier",
"@vue/prettier/@typescript-eslint"
```
If that's ok with you I think it's good to follow this default configuration offered by the cli, without adding the `plugin:vue/strongly-recommended` and `plugin:vue/recommended`, both because it's good practice to follow the CLI setup (as it is thoroughly tested and consolidated) and because using multiple recommended plugins could end up causing us issues (let's say if they conflict with the recommended stuff for typescript). What do you think?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -85,34 +86,36 @@ Before submitting your change, you need to make sure that they do not break exis
 #### 6: Lint and Prettify code
 
 Any code you submit must:
-  * Not have any errors detectable by [ESLint](https://eslint.org/). ESLint enables us to easily detect and solve common problems in our code.  
-  * Be formatted with [Prettier](https://prettier.io/). We use prettier to have a consistent and automatic code formatting without having to memorize and follow a style guide.
```
[the remark this answers is not in the document]
Comment (aleesaan) [reply CONTRIBUTING.md:89]: I'm not sure how it happened, I'll revert this change! We need to mention both prettier and eslint, they have different functionality (has explained in these paragraphs 😅).

File: .eslintrc.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  parserOptions: {
+    parser: "@typescript-eslint/parser",
+    sourceType: "module",
+    ecmaVersion: 2020
+  },
+  rules: {
+    "no-console": process.env.NODE_ENV === "production" ? "warn" : "off"
```
[the remark this answers is not in the document]
Comment (aleesaan) [reply .eslintrc.js:13]: Same as for the other answer, I used the default configs from the vue cli. I don't think we need the extra rules since we use the recommended ones, and I think they're working pretty great! Unless you see the edge cases that needed these rules, I'd say to leave them out. What do you think?
As for your question about the eslint in production, yes! You can already test it by adding a `console.log` somewhere and changing the `lint` script to `vue-cli-service lint --mode production`, which will then throw a warning.

File: .prettierrc.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (SamLubbers) [reply .prettierrc.json]: * good to know. Let's keep it that way
* yes I previously thought that it always had to be 80 but I learned this week that 120 is the new standard. This works much better for the html in our vue files
* yes good to use the default values.

File: .eslintrc.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (SamLubbers) [reply .eslintrc.js]: sounds good to use the recommended CLI standards. If our code really needs more quality enforcement we can add more rules in the future :+1:

File: .eslintrc.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (SamLubbers) [reply .eslintrc.js:4]: no me neither, so let's go with the CLI standard.

File: .prettierrc.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (aleesaan) [reply .prettierrc.json]: Ah I actually dropped the `120` in the end 😅  shall I bring it back? I do think 80 is quite short!

[the remark this answers is not in the document]
Comment (SamLubbers) [reply .prettierrc.json]: Yes you can bring it back 😂

File: .eslintrc.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (SamLubbers) [reply .eslintrc.js:19]: - good to know about the production flag on eslint. Then we can easily test that before deploying
- I think that custom configuration I added was to solve a conflict between prettier and the Vue recommended rules. Since we don't have that rule anymore we can keep it out.
- for prettier, I think it's still useful to add the warning so that it tells you what code isn't formatted right. Or maybe that isn't necessary if prettier autoformats the code. What do you think?

[the remark this answers is not in the document]
Comment (aleesaan) [reply .eslintrc.js:19]: Yep it's not needed! Prettier will format automatically based on our lint rules so we don't need a warning.

**Review by SamLubbers — APPROVED** [review]
@aleesaan I finally managed to finsih reviewing it (on my phone, on the train on my way to the antiracism protest in The Hague).

The code looks good to merge. The js to ts changes are actually very minor; most were eslint fixes.

In future pull requests I would first make a request with all the lining fixes and then all the typescript changes. Then it's easier to see what changed.

Thanks again for this big improvement! 💯
